### ISSUE TITLE
Add a New Selection

### ISSUE THREAD

**Issue by SouIchimiya** [IfcOpenShell/IfcOpenShell#3459]
Hi, I'm Japanese student and studying BlenderBIM.

I am currently considering adding Uniclass2015 as an option in the IFC selection menu to define a wider range of data. So, if I wanted to add another tab as shown in the image below, can it be implemented using Blender's built-in features? If Python scripting is required, which code should be modified? Honestly, I'm not very proficient in programming (I only know how to read it to some extent), so I would appreciate any advice on alternative implementation methods if you have any ideas.

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/129352032/39fa7511-e0e8-4b5f-89aa-8f63caaefa76)

And finally, if there is already a similar question that I asked in the past and if that thread exists, I would appreciate it if you could kindly provide me with the page.

**Comment by Moult** [comment]
Adding classification based selection is a great idea!

Adding another UI there to assign the classification at the same time as assigning the class is useful, but some thought needs to go into exactly what the UI will be. A dropdown is not practical as classifications typically have hundreds of classification references if not more. Currently we use a hierarchical tree in the IFC classification panel. Have you tested that out?

**Comment by SouIchimiya** [comment]
I greatly appreciate your prompt response. 

In fact, I have been independently working on the mapping of IFC and Uniclass2015. The content involves categorizing from Uniclass_Ss to IFC Wall Type Enum. (The usefulness and accuracy of the mapping results may not be certain, but if I can at least present the methodology, I believe I can leave the rest to experts who are more knowledgeable than me.) 

By specifying the IFC Class, I would like to narrow down the selection to some extent. Specifically, I am considering incorporating Uniclass2015 into further substructures so that when selecting an IFC Element, I can also select the associated entities. (I apologize if my explanation is unclear.)

**Comment by Moult** [comment]
Ah yes in that case a dropdown would be good ... as long as someone has sat down and done the _huge_ amount of work to map them!

Not to mention Uniclass has updates several times a year so someone has to maintain this mapping! It's a huge task and kudos if you want to go through it.

According to the work done in facility management in IFC I believe the Uniclass Ss table should not map to IFC (physical) elements. Roughly speaking:

 - Uniclass Co maps to IfcProjects
 - Uniclass En maps to IfcFacilities
 - Uniclass SL maps to any non-facility space, storey, zone, etc
 - Uniclass Ss maps to IfcSystems
 - Uniclass En maps to any generic physical element (walls, slabs, coverings, plates, etc)
 - Uniclass Pr maps to any maintainable product / equipment (pumps, chillers, etc ... these are already identified in the IFC equipment asset handover (the name keeps on changing) project) and are almost the same as what was in COBie years ago.

So assuming you have a mapping, I would publish this mapping in some form that software can easily use and keep up to date with (e.g. a JSON dictionary) and then creating a filtered dropdown in that case is relatively trivial and I'd be more than happy to show you.

BTW in Blender you can right click anywhere in the interface and press "Edit Source" to edit the code in that part of the UI in Blender's text editor.

**Comment by SouIchimiya** [comment]
I apologize for the delayed response.

When considering the practical use of BlenderBIM, I believe that a more detailed definition of objects becomes necessary. I think that relying solely on IFC Class and Psets is insufficient for accurate architectural information, and I hope that Uniclass_Ss can fill that gap. (Since Uniclass_En and Uniclass_Ss have a one-to-many correspondence, mapping with general physical elements should also be possible, right?)

Indeed, mapping with Uniclass requires a significant amount of effort, and it may be beyond the capabilities of an individual. However, if someone discovers that potential, I believe others would also raise their voices to give it a try. (It would be great if they did!)

If my thinking is incorrect, please let me know.

**Comment by Moult** [comment]
Absolutely, local classification standards are used for a reason (although I find personally that IFC Class and Predefined Type is far more descriptive than people give it credit for). I'm just saying the Uniclass Ss table isn't exactly meant for single objects. It's meant to be used to describe a **collection of products** working together as a system. Therefore you shouldn't map a single "IfcWall" to a system. Instead an IfcSystem, IfcElementAssembly, or IfcGroup, an aggregated entity or similar might be classified using Uniclass Ss.

From https://www.thenbs.com/knowledge/uniclass-tables-explained :

> The Systems table classifies **collections of products** brought together to operate as systems, in order to provide a common purpose or solution. Systems can vary in complexity, from just a few products working in combination in a space/ location, to vast networks of products working together as a system over large areas. For example, a rail signal system is made up of signals, power cables, data distribution, detection and warning equipment, posts, lighting etc.

**Comment by SouIchimiya** [comment]
You're right, it may not be appropriate to associate Uniclass Ss with individual objects. In that case, it seems like my task would involve mapping IfcSystem to Uniclass Ss.

By the way, in BlenderBIM, what features or methods are available for grouping multiple objects using IfcGroup?

**Comment by Moult** [comment]
Exactly. And this is where you might become a little stuck if you want to focus on Uniclass Ss. IfcSystems have minimal classifications (e.g. you can have a distribution system, or a building system, or a circuit, but not much more). So most of the identification (e.g. this is a Hot water system) comes from the IfcSystem name or description, which are not standardised and vary from country / project.

BBIM supports IfcGroup so you can group any product in a group.

**Comment by Moult** [comment]
Totally agree, would you perhaps consider mapping the Pr and EF tables instead to individual IFC classes? Those would be the most appropriate.

https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/util/fm.py#L78-L113 these classes would map to Uniclass Pr and any other IfcElement class that isn't part of that list would link to an EF table.
